river/neighbors/knn_classifier.py

```diff
@@ -1,176 +1,193 @@
+        Returns:
+            self
+        """
+        # Only add the class y to known classes if we actually add the point!
+        if self.nn.update((x, y), n_neighbors=self.n_neighbors):
+            self.classes.add(y)
 
-        Parameters
-        ----------
-        x
-            A dictionary of features.
-        y
-            The class label.
+        # Ensure classes known to instance reflect window
+        self._run_class_cleanup()
+        return self
 
-        Returns
-        -------
-            self
+    def _run_class_cleanup(self):
+        """
+        Helper function to run class cleanup, accounting for _cleanup_counter.
+        """
+        # clean up classes every cleanup_every steps
+        if self.cleanup_every:
+            self._cleanup_counter -= 1
+            if self._cleanup_counter == 0:
+                self.clean_up_classes()
+                self._cleanup_counter = self.cleanup_every
 
-        Notes
-        -----
-        For the K-Nearest Neighbors Classifier, fitting the model is the
-        equivalent of inserting the newer samples in the observed window,
-        and if the size_limit is reached, removing older results.
+        return self
 
+    def predict_one(self, x: dict):
+        """Predict the label of a set of features `x`.
+        Parameters:
+            x: A dictionary of features.
+        Returns:
+            The neighbors
         """
+        return self.predict_proba_one(x)
 
-        self.classes_.add(y)
-        x_arr = dict2numpy(x)
+    def predict_proba_one(self, x):
+        """Predict the class of a set of features `x`.
+        Parameters:
+            x: A dictionary of features.
+        Returns:
+            Lookup (dict) of classes and probability predictions (normalized)
+        """
+        nearest = self.nn.find_nearest((x, None), n_neighbors=self.n_neighbors)
 
-        self.data_window.append(x_arr, y)
+        # Default prediction for every class we know is 0.
+        # If class_cleanup is false this can include classes not in window
+        y_pred = {c: 0.0 for c in self.classes}
 
-        return self
+        # No nearest points? Return the default.
+        if not nearest:
+            return y_pred
 
-    def predict_proba_one(self, x):
-        """Predict the probability of each label for a dictionary of features `x`.
+        # If the closest is an exact match AND has a class, return it
+        if nearest[0][-1] == 0 and nearest[0][0][1] is not None:
 
-        Parameters
-        ----------
-        x
-            A dictionary of features.
+            # Update the class in our prediction from 0 to 1, 100% certain!
+            y_pred[nearest[0][0][1]] = 1.0
+            return y_pred
 
-        Returns
-        -------
-        proba
-            A dictionary which associates a probability which each label.
+        for neighbor in nearest:
+            distance = neighbor[-1]
+            y = neighbor[0][1]
```

Comment:
You can also do `(x, y), distance = neighbor`, which might be more readable, but it's your call

Reply:
I originally did the above because we couldn't guarantee the number of items to unwrap. I don't think that's the case anymore because it would be represented in x so happy to change it here, although it could lead to issues I can't anticipate now.